src/sync-external-store--stateful-store/store.ts

```diff
@@ -0,0 +1,44 @@
+import { SyncExternalStore } from "../sync-external-store/index.js";
+import {
+  CreateErrorSnapshot,
+  CreateLoadingSnapshot,
+  CreateSuccessSnapshot,
+  createErrorSnapshot,
+  createLoadingSnapshot,
+  createSuccessSnapshot,
+} from "./create-snapshot.js";
+import { StoreState } from "./store-item.js";
+
+type UpdateStoreParams<TStore extends StoreState, TParams = void, TResponse = void> = {
+  key: keyof TStore;
+  params?: TParams;
+  service: (params: TParams) => Promise<TResponse>;
+  updateSnapshot?: {
+    loading: CreateLoadingSnapshot;
+    success: CreateSuccessSnapshot;
+    error: CreateErrorSnapshot;
+  };
+};
+
+export abstract class StatefulSyncExternalStore<TStore extends StoreState> extends SyncExternalStore<TStore> {
+  async updateStore<TParams, TResponse>(params: UpdateStoreParams<TStore, TParams, TResponse>) {
```

Comment:
Okay playing with this the code I got to was something like this:

```
  public getOne(id: string) {
    return this.updateStore({
      key: id,
      service: (id: string) => Promise.resolve("hello" + id),
    });
  }
  ```
  
which is a bit weird that `id` is reused soo many times. Also the only thing that changes here after initialization is id and params, service and updateSnapshot should be constants through the lifetime of the class.

I think there's some code reorganization needed to split these apart

Reply:
Maybe this is actually a good use of currying weirdly...

Reply:
Here's the dumb non-currying version of some sample code which would be nice to be able to write (if we can get it be less lines of code anyway)

```
class SomeStore extends StatefulSyncExternalStore<Record<string, StoreItem<string>>> {
  private getOneService = this.createUpdateStore({
    service: (id: string) => Promise.resolve("hello" + id),
  });
  public getOne(id: string) {
    return this.getOneService({
      key: "id",
      params: id,
    });
  }
}

function run() {
  const d = new SomeStore({});
  d.getOne("id-#1");
}
```

Reply:
Document on currying
https://www.freecodecamp.org/news/playing-around-with-closures-currying-and-cool-abstractions/